log/log.go

```diff
@@ -0,0 +1,134 @@
+package log
+
+import (
+	"context"
+	"fmt"
+	tmlog "github.com/tendermint/tendermint/libs/log"
+)
+
+// Logger is a simple interface to log at three log levels with additional formatting methods for convenience
+type Logger interface {
+	Debug(msg string)
+	Debugf(format string, a ...any)
+	Info(msg string)
+	Infof(format string, a ...any)
+	Error(msg string)
+	Errorf(format string, a ...any)
+}
+
+var (
+	defaultLogger = logWrapper{tmlog.NewNopLogger()}
+	frozen        bool
+)
+
+// Setup sets the logger that the application should use. The default is a nop logger, i.e. all logs are discarded.
+// Panics if called more than once without calling Reset first.
+func Setup(logger tmlog.Logger) {
+	if frozen {
+		panic("logger was already set")
+	}
+
+	defaultLogger = logWrapper{logger}
+	frozen = true
+}
+
+// Reset returns the logger state to the default nop logger and enables Setup to be called again.
+func Reset() {
+	defaultLogger = logWrapper{tmlog.NewNopLogger()}
+	frozen = false
+}
+
+// Debug logs the given msg at log level DEBUG
+func Debug(msg string) {
+	defaultLogger.Debug(msg)
+}
+
+// Debugf logs the given formatted msg at log level DEBUG
+func Debugf(format string, a ...any) {
+	defaultLogger.Debugf(format, a...)
+}
+
+// Info logs the given msg at log level INFO
+func Info(msg string) {
+	defaultLogger.Info(msg)
+}
+
+// Infof logs the given formatted msg at log level INFO
+func Infof(format string, a ...any) {
+	defaultLogger.Infof(format, a...)
+}
+
+// Error logs the given msg at log level ERROR
+func Error(msg string) {
+	defaultLogger.Error(msg)
+}
+
+// Errorf logs the given formatted msg at log level ERROR
+func Errorf(format string, a ...any) {
+	defaultLogger.Errorf(format, a...)
+}
+
+type key int
+
+var logKey key
```

Comment:
how is this set?

Reply:
doesn't need to be set, context uses the type itself to differentiate. Therefore recommended practice is to derive a new private key type from `int` so no other package can have any key collision. Note that only the topmost match with a key is returned, therefore the keyval slice is always appended and attached to the newly created context layer